### ISSUE TITLE
Properly calculate balances (free, reserved, locked)

### ISSUE THREAD

**Issue by jacogr** [polkadot-js/api#876]
I believe we should drop the actual nominator balances stuff completely, they are available now via the staking interfaces and doesn't need to be calculated. (Effectively this is basically almost ignore what is there in terms of balance retrieval and redo)

Once properly retrieved (including locked, available, free, reserved), can pull this into the UI and drop any recent calculations there.

cc @Tbaut  - it could help if you drop some of the pointers from your travels in the last couple of days
cc @kwingram25 - as discussed

**Comment by kwingram25** [comment]
What is the new staking interface? Is there one call to retrieve all of this now?

**Comment by jacogr** [comment]
No, not new - well, it is new since the time this code was originally done, but just taking into the account the actual bonded/locked balances for staking.

**Comment by Tbaut** [comment]
From my understanding and tests:
Available balance (that can be transferred) is:
- available = `vestedBalance - maxLocks`
- `vestedBalance` is `balances_vesting.perBlock`x`currentBlock`+`( balances_freeBalance - balances_vesting.offset )` 
- `maxLocks` is the maximum amount from `balances_locks.unlocking[]` given that `balances_locks.unlocking.era` < current era.

The [tests in balances srml module](https://github.com/paritytech/substrate/blob/master/srml/balances/src/tests.rs#L44) is a good way to understand how things work.

In plain English:

Vesting is a way to slowly let users get access to their funds from a certain point in time. If I have a `freeBalance` of 20 Units, a vesting offset of 30 blocks, and vesting per block 1 Unit, I'll start getting access to my money starting block 10, and have access to all the funds by the block 30 (the offset). Vesting can only be made in the genesis (changing the `chain_spec`).

Now from that vested funds, some can be locked (e.g when bonding to a controller account in order to validate).

**Comment by joepetrowski** [comment]
I can help you with the Balances module and Currency trait. Some things that I think were unclear from @Tbaut :

> `available = vestedBalance - maxLocks`

Available is `free_balance - locks`. (The idea of "max" locks doesn't make much sense since they apply over each other. I.e. reserves add, locks overlay. That said, when checking liquidity restrictions in `ensure_account_can_withdraw`, we do iterate through locks just in case, but `locks.len()` should always be `0 || 1`.)

> If I have a freeBalance of 20 Units, a vesting offset of 30 blocks, and vesting per block 1 Unit, I'll start getting access to my money starting block 10, and have access to all the funds by the block 30 (the offset).

Actually it is simpler than that. You start gaining access to your funds from block zero. You just have a balance locked at genesis called `offset` (I think this is a poor variable name choice, potentially leading to the confusion about when you start gaining access, and as Thibaut said, can only be made at genesis using `chain_spec`) and then a fixed amount that you gain access to each block called `per_block`.

There isn't any inherent logic in calculating the end of the vest, when you have all your funds. If you don't want a user to have access to `x` funds until block `n`, you'd have to do the calculation yourself and input the correct genesis config. Since it's a simple linear de-vest, it should be simple. Would actually be interesting (and not too difficult) to add a few vesting schedules in an `enum` and let you add a variant to the config (e.g. vesting with cliffs, vesting starting after block `n`, non-linear devesting).

```
/// Struct to encode the vesting schedule of an individual account.
#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(Debug))]
pub struct VestingSchedule<Balance> {
	/// Locked amount at genesis.
	pub offset: Balance,
	/// Amount that gets unlocked every block from genesis.
	pub per_block: Balance,
}

impl<Balance: SimpleArithmetic + Copy + As<u64>> VestingSchedule<Balance> {
	/// Amount locked at block `n`.
	pub fn locked_at<BlockNumber: As<u64>>(&self, n: BlockNumber) -> Balance {
		if let Some(x) = Balance::sa(n.as_()).checked_mul(&self.per_block) {
			self.offset.max(x) - x
		} else {
			Zero::zero()
		}
	}
}
```
A little example:
```
offset=100 // locked amount at genesis
per_block=5

block   locked
0		100 (max(100,0*5) - 0*5)
10		50 (max(100,10*5) - 10*5)
20		0 (max(100,20*5) - 20*5)
50		0 (max(100,50*5) - 50*5)
```

**Comment by xlc** [comment]
I can see a fundamental problem here that why do we need to implement this balance logic multiple times? One in Rust, one in JS, people are going to need Go, Java, Python SDK later and do they need to implement this complicated logic in each languages again?

We really need a query API like described here https://github.com/paritytech/substrate/issues/2214#issuecomment-480217815

**Comment by jacogr** [comment]
@kwingram25 Anything left before we close this?

**Comment by kwingram25** [comment]
I don't think so, @Tbaut ?

**Comment by Tbaut** [comment]
not that I see :)

**Comment by Anukshuk** [comment]
Sorry i am not sure if i am at the right place. I  need help with my locked polkadot. Please help
